File: app/src/main/java/pl/gocards/ui/ExportImportDbUtil.java
package pl.gocards.ui;

import static io.reactivex.rxjava3.internal.functions.Functions.EMPTY_ACTION;

import android.annotation.SuppressLint;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.FileUtils;
import android.provider.OpenableColumns;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.channels.FileChannel;
import java.util.Objects;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import pl.gocards.room.util.DbUtil;
import pl.gocards.db.deck.AppDeckDbUtil;
import pl.gocards.db.storage.DatabaseException;
import pl.gocards.ui.main.MainActivity;
import pl.gocards.util.ExceptionHandler;

/**
 * @author Grzegorz Ziemski
 */
public class ExportImportDbUtil {

    private static final String TAG = "ExportImportDbUtil";

    protected String path;

    @NonNull
    protected final AppCompatActivity activity;

    @NonNull
    protected final ActivityResultLauncher<String> exportDbLauncher;

    @NonNull
    protected final ActivityResultLauncher<String[]> importDbLauncher;

    public ExportImportDbUtil(@NonNull AppCompatActivity activity) {
        this.activity = activity;
        exportDbLauncher = initExportDbLauncher();
        importDbLauncher = initImportDbLauncher();
    }

    /**
     * D_R_12 Export database
     */
    @NonNull
    protected ActivityResultLauncher<String> initExportDbLauncher() {
        return activity.registerForActivityResult(
                new ActivityResultContracts.CreateDocument("application/vnd.sqlite3"),
                exportedDbUri -> {
                    if (exportedDbUri != null) exportDb(exportedDbUri, path);
                }
        );
    }

    /**
     * D_C_13 Import database
     */
    @NonNull
    protected ActivityResultLauncher<String[]> initImportDbLauncher() {
        return activity.registerForActivityResult(
                new ActivityResultContracts.OpenDocument(),
                importedDbUri -> {
                    if (importedDbUri != null)
                        importDb(path, importedDbUri);
                }
        );
    }

    /**
     * D_R_12 Export database
     */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    @SuppressLint("CheckResult")
    protected void exportDb(Uri exportToUri, @NonNull String dbPath) {
        Completable.fromRunnable(() -> {
                    try {
                        AppDeckDbUtil
                                .getInstance(activity)
                                .flushDatabase(dbPath);
                    } catch (DatabaseException e) {
                        throw new RuntimeException(e);
                    }

                    try (OutputStream out = activity
                            .getContentResolver()
                            .openOutputStream(exportToUri)) {
                        try (FileInputStream in = new FileInputStream(dbPath)) {
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                                FileUtils.copy(in, Objects.requireNonNull(out));
                            } else {
                                FileChannel inChannel = in.getChannel();
                                FileChannel outChannel = ((FileOutputStream) Objects.requireNonNull(out)).getChannel();
                                inChannel.transferTo(0, inChannel.size(), outChannel);
                            }
                        }
                    } catch (Exception e) {
                        onErrorExportDb(e);
                    }
                })
                .subscribeOn(Schedulers.io())
                .subscribe(EMPTY_ACTION, this::onErrorExportDb);
    }

    protected void onErrorExportDb(@NonNull Throwable e) {
        getExceptionHandler().handleException(
                e, activity,
                this.getClass().getSimpleName(),
                "Error while exporting DB."
        );
    }

    /**
     * D_C_13 Import database
     */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    @SuppressLint({"Range", "CheckResult"})
    protected void importDb(@NonNull String importToFolder, Uri importedDbUri) {
        Completable.fromRunnable(() -> {
                    AppDeckDbUtil folderDeckDbUtil = AppDeckDbUtil
                            .getInstance(activity.getApplicationContext());
                    try {
                        String deckDbPath = null;
                        try (Cursor cursor = activity.getContentResolver()
                                .query(importedDbUri, null, null, null)) {
                            if (cursor != null && cursor.moveToFirst()) {
                                deckDbPath = cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME));
                                deckDbPath = folderDeckDbUtil.findFreePath(importToFolder, deckDbPath);
                            }
                        }
                        InputStream in = activity.getContentResolver().openInputStream(importedDbUri);
                        Objects.requireNonNull(in);
                        FileOutputStream out = new FileOutputStream(deckDbPath);
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                            FileUtils.copy(in, out);
                        } else {
                            FileChannel inChannel = ((FileInputStream) in).getChannel();
                            FileChannel outChannel = out.getChannel();
                            inChannel.transferTo(0, inChannel.size(), outChannel);
                        }
                        in.close();
                        out.close();
                        if (activity instanceof MainActivity) {
                            ((MainActivity) activity).refreshItems();
                        }
                    } catch (IOException e) {
                        onErrorImportDb(e);
                    }
                })
                .subscribeOn(Schedulers.io())
                .subscribe(EMPTY_ACTION, this::onErrorImportDb);
    }

    protected void onErrorImportDb(@NonNull Throwable e) {
        getExceptionHandler().handleException(
                e, activity, TAG,
                "Error while importing DB."
        );
    }

    /**
     * D_R_12 Export database
     */
    public void launchExportDb(@NonNull String dbPath) {
        path = dbPath;
        exportDbLauncher.launch(DbUtil.addDbExtension(getDeckName(dbPath)));
    }

    /**
     * D_C_13 Import database
     */
    public void launchImportDb(String importToFolder) {
        path = importToFolder;
        importDbLauncher.launch(new String[]{
                "application/vnd.sqlite3",
                "application/octet-stream"
        });
    }

    @NonNull
    protected String getDeckName(@NonNull String dbPath) {
        return AppDeckDbUtil.getDeckName(dbPath);
    }

    @NonNull
    protected ExceptionHandler getExceptionHandler() {
        return ExceptionHandler.getInstance();
    }
}

File: app/src/main/java/pl/gocards/ui/decks/kt/decks/service/ExportImportDbUtil.kt
package pl.gocards.ui.decks.kt.decks.service

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.FileUtils
import android.provider.OpenableColumns
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.annotation.UiThread
import pl.gocards.db.deck.AppDeckDbUtil
import pl.gocards.db.storage.DatabaseException
import pl.gocards.room.util.DbUtil
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.Objects

/**
 * @author Grzegorz Ziemski
 */
open class ExportImportDbUtil(protected var context: Context) {
    
    protected lateinit var path: String

    /**
     * D_R_12 Export database
     */
    protected fun launchExportDb(
        exportDbLauncher: ActivityResultLauncher<String>,
        dbPath: String
    ) {
        path = dbPath
        exportDbLauncher.launch(DbUtil.addDbExtension(getDeckName(dbPath)))
    }

    /**
     * D_C_13 Import database
     */
    protected fun launchImportDb(
        importDbLauncher: ActivityResultLauncher<Array<String>>,
        importToFolder: String
    ) {
        path = importToFolder
        importDbLauncher.launch(
            arrayOf(
                "application/vnd.sqlite3",
                "application/octet-stream"
            )
        )
    }

    /**
     * D_R_12 Export database
     */
    @SuppressLint("CheckResult")
    protected open fun exportDb(
        exportToUri: Uri,
        dbPath: String
    ) {
        flushDatabase(dbPath)

        context
            .contentResolver
            .openOutputStream(exportToUri)
            .use { output ->
                FileInputStream(dbPath)
                    .use { input ->
                        copyData(input, output!!)
                    }
            }
    }

    private fun flushDatabase(dbPath: String) {
        try {
            getAppDeckDbUtil()
                .flushDatabase(dbPath)
        } catch (e: DatabaseException) {
            throw RuntimeException(e)
        }
    }

    private fun copyData(input: FileInputStream, output: OutputStream) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            FileUtils.copy(input, Objects.requireNonNull(output))
        } else {
            val inChannel = input.channel
            val outChannel = (output as FileOutputStream).channel
            inChannel.transferTo(0, inChannel.size(), outChannel)
        }
    }

    /**
     * D_C_13 Import database
     */
    @SuppressLint("Range", "CheckResult")
    protected open fun importDb(
        importToFolder: String,
        importedDbUri: Uri,
        onSuccess: (deckDbPath: String) -> Unit
    ) {
        try {
            var deckDbPath: String = getDeckDbPath(importedDbUri)!!
            deckDbPath = findFreePath(importToFolder, deckDbPath)

            context.contentResolver.openInputStream(importedDbUri)!!
                .use { input ->
                    FileOutputStream(deckDbPath)
                        .use { output ->
                            copyData(input, output)
                        }
                }

            onSuccess(deckDbPath)
        } catch (e: IOException) {
            onErrorImportDb(e)
        }
    }

    protected open fun onErrorImportDb(e: Throwable) {
        e.printStackTrace()
    }

    private fun copyData(input: InputStream, output: FileOutputStream) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            FileUtils.copy(input, output)
        } else {
            val inChannel = (input as FileInputStream?)!!.channel
            val outChannel = output.channel
            inChannel.transferTo(0, inChannel.size(), outChannel)
        }
    }

    @SuppressLint("Range")
    private fun getDeckDbPath(importedDbUri: Uri): String? {
        context
            .contentResolver
            .query(importedDbUri, null, null, null)
            .use { cursor ->
                if (cursor != null && cursor.moveToFirst()) {
                    return cursor.getString(cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME))
                }
            }

        return null
    }

    private fun findFreePath(
        importToFolder: String,
        deckDbPath: String
    ): String {
        return getAppDeckDbUtil().findFreePath(importToFolder, deckDbPath)
    }

    protected fun getDeckName(dbPath: String): String {
        return AppDeckDbUtil.getDeckName(dbPath)
    }

    private fun getAppDeckDbUtil(): AppDeckDbUtil {
        return AppDeckDbUtil.getInstance(context)
    }

    @UiThread
    protected fun showShortToastMessage(text: String) {
        Toast.makeText(context, text, Toast.LENGTH_SHORT).show()
    }
}
